### PR TITLE
Adding step6 tutorial.

### PR DESCRIPTION
**Pull request by jmark**
**_Describe your changes here:_**

Added another tutorial step. The wiki page can be found in https://github.com/DLR-AMR/t8code/wiki/Step-6-Computing-stencils .

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

- [ ] The author added a BSD statement to `doc/` (or already has one)
- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)
- [ ] New source/header files are properly added to the Makefiles
- [ ] New Datatypes are added to t8indent_custom_datatypes.txt
- [ ] The reviewer executed the new code features at least once and checked the results manually
- [ ] The code is covered in an existing or new test case
- [ ] New tests use the Google Test framework
- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

### CODE REVIEW

File: tutorials/general/t8_step6_stencil.cxx [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+  /* Destroy the wrapper array. This will not free the data memory since we used sc_array_new_data. */
+  sc_array_destroy (sc_array_wrapper);
+}
+
+/* Write the forest as vtu and also write the element's volumes in the file.
+ * 
+ * t8code supports writing element based data to vtu as long as its stored
+ * as doubles. Each of the data fields to write has to be provided in its own
+ * array of length num_local_elements.
+ * We support two types: T8_VTK_SCALAR - One double per element
```
Comment (chiaraMaHe) [inline tutorials/general/t8_step6_stencil.cxx:288]: 1 double or Three doubles (I think the second one is better)

File: tutorials/general/t8_step6_stencil.cxx [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with t8code; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/* See also: https://github.com/holke/t8code/wiki/Step-6---Compute stencils.
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:23]: Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    {0.0, 0.0, 0.0},              /* Midpoints of the sphere. */
+    0.5,                        /* Refine if inside this radius. */
+    0.7                         /* Coarsen if outside this radius. */
+  };
+  /* Start with a uniform forest. */
+  t8_forest_t         forest =
+    t8_forest_new_uniform (cmesh, scheme, level, 0, comm);
+  t8_forest_t         forest_apbg;
+
+  /* Adapt, partition, balance and create ghost elements all in the same step. */
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:81]: Reformulated it. Done.

[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:81]: Added a reference. Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+  t8_locidx_t         itree, num_local_trees;
+  t8_locidx_t         current_index;
+  t8_locidx_t         ielement, num_elements_in_tree;
+  t8_eclass_t         tree_class;
+  t8_eclass_scheme_c *eclass_scheme;
+  const t8_element_t *element;
+
+  double verts[4][3] = {0};
+
+  /* Check that forest is a committed, that is valid and usable, forest. */
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:109]: Shortened it. Details were already explained in previous steps. Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+  T8_ASSERT (t8_forest_is_committed (forest));
+
+  /* Get the number of local elements of forest. */
+  num_local_elements = t8_forest_get_local_num_elements (forest);
+  /* Get the number of ghost elements of forest. */
+  num_ghost_elements = t8_forest_get_num_ghosts (forest);
+
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:116]: Comment was added. Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+{
+  int                 level;
+  double              midpoint[3];
+  double              dx, dy;
+  double              volume;
+  double              height;
+  double              schlieren;
+  double              curvature;
+};
+
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:64]: Added. Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+  t8_forest_set_user_data (forest_apbg, &adapt_data);
+  t8_forest_set_adapt (forest_apbg, forest, t8_step3_adapt_callback, 0);
+  t8_forest_set_partition (forest_apbg, NULL, 0);
+  t8_forest_set_balance (forest_apbg, NULL, 0);
+  t8_forest_set_ghost (forest_apbg, 1, T8_GHOST_FACES);
+  t8_forest_commit (forest_apbg);
+
+  return forest_apbg;
+}
+
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:92]: Added. Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+      const double r = sqrt(x*x + y*y)*20.0; // scaled radius
+
+      /* Some 'interesting' height function. */
+      edat->height = sin(2.0*r)/r;
+    }
+  }
+
+  return element_data;
+}
+
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:159]: Added. Done.

File: tutorials/general/t8_step6_stencil.cxx (hunk 1 above)
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:288]: Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+T8_EXTERN_C_BEGIN ();
+
+/* The data that we want to store for each element.
+ * In this example we want to store the element's level and volume. */
+struct data_per_element
+{
+  int                 level;
+  double              midpoint[3];
+  double              dx, dy;
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:58]: Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+    for (ielement = 0; ielement < num_elements_in_tree; ++ielement, ++current_index) {
+      element = t8_forest_get_element_in_tree (forest, itree, ielement);
+
+      stencil[1][1] = element_data[current_index].height;
+      dx[1] = element_data[current_index].dx;
+      dy[1] = element_data[current_index].dy;
+
+      num_faces = eclass_scheme->t8_element_num_faces (element);
+      for (iface = 0; iface < num_faces; iface++) {
+
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:202]: Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+  /* Get the number of local elements of forest. */
+  num_local_elements = t8_forest_get_local_num_elements (forest);
+  /* Get the number of ghost elements of forest. */
+  num_ghost_elements = t8_forest_get_num_ghosts (forest);
+
+  element_data = T8_ALLOC (struct data_per_element, num_local_elements + num_ghost_elements);
+
+  /* Get the number of trees that have elements of this process. */
+  num_local_trees = t8_forest_get_num_local_trees (forest);
+
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:121]: Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+  T8_ASSERT (t8_forest_is_committed (forest));
+
+  /* Get the number of trees that have elements of this process. */
+  num_local_trees = t8_forest_get_num_local_trees (forest);
+
+  double stencil[3][3] = {0};
+  double dx[3] = {0};
+  double dy[3] = {0};
+
+  for (itree = 0, current_index = 0; itree < num_local_trees; ++itree) {
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:187]: Done.

File: tutorials/general/t8_step6_stencil.cxx [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+            stencil[1][2] = height;
+            dy[2] = element_data[neighids[0]].dy;
+            break;
+        }
+
+        T8_FREE(neighbors);
+        T8_FREE(dual_faces);
+        T8_FREE(neighids);
+      }
+
```
[the remark this answers is not in the document]
Comment (jmark) [reply tutorials/general/t8_step6_stencil.cxx:255]: Added more info in the wiki. Done.

**Review by chiaraMaHe — APPROVED** [review]

**Pull request merged**